BLD: Revamp BLAS/LAPACK G77 ABI wrappers and fix PROPACK on x86

thalassemia opened the pull request:
#### Reference issue
<!--Example: Closes gh-WXYZ.-->
Closes gh-15108. Includes all the changes in gh-18354.

#### What does this implement/fix?
Taking the lessons I learned from implementing Accelerate support, I rewrote the G77 BLAS/LAPACK ABI wrappers entirely in C. These wrappers should be compatible with any Fortran or C code that expects the standard C99 `_Complex` type. To make `cython_blas` and `cython_lapack` work despite Cython C using struct complex types, I let `scipy/linalg/_generate_pyx.py` continue to generate an intermediate Fortran wrapper between the Cython C and the G77 wrapper C. This PR includes a temporary commit (c69d03b) to make use of changes to the PROPACK repo that would ideally be merged concomitantly with this one. 

With these changes, I believe I have fixed the segfaults and other issues previously seen with PROPACK on 32-bit machines, allowing me to re-enable all the relevant tests and 32-bit support for PROPACK in general.

#### Additional information
I had to tweak one of the tolerances in the (currently skipped) `test_svds_parameter_tol` test in `test_svds.py` for it to pass on my machine. I am not sure why there is a lower bound on the output error, but that was also causing some test failures on my 32-bit Linux VM.

h-vetinari committed:
Revert "MAINT: add `SCIPY_USE_PROPACK` env variable (#16361)"

This reverts commit c73e0883501a21cfaf7220a8118a6155857b5829.

thalassemia committed:
BLD: C wrappers for G77 ABI BLAS/LAPACK

thalassemia committed:
BLD: Temp use PROPACK fork with fixed G77 wrappers

thalassemia committed:
TST: Bring back all PROPACK tests

thalassemia committed:
TST: Increase LOBPCG parameter tol

thalassemia committed:
TST: Remove SCIPY_USE_PROPACK from CI

thalassemia committed:
TST: Try enabling 32-bit PROPACK tests

thalassemia commented:
I've been noticing sporadic failures on the Windows `cp311 (build sdist + wheel), full, no pythran` test for a while now, and I do not know what is causing it.

Per comments on #19816, I think it would be worth taking some time to document and refactor `scipy/linalg/_generate_pyx.py`, from which I copied a lot of code when writing `scipy/linalg/_generate_blas_wrapper.py` in that PR. I'm not sure whether to create a separate PR for that or do it here.

thalassemia committed:
TST: LOBPCG expect accuracy warning

thalassemia committed:
MAINT: Fix int type and missing & in G77 wrappers

thalassemia committed:
TST: Enable np.complex128 PROPACK test on Windows

thalassemia committed:
MAINT: Remove unused type: ignore

h-vetinari commented:
> With these changes, I believe I have fixed the segfaults and other issues previously seen with PROPACK on 32-bit machines

That would indeed be amazing! 🤩 

Just for clarity, the segfaults not only occurred on 32-bit machines, but also 64-bit ones, see https://github.com/conda-forge/scipy-feedstock/pull/200. It's mainly on windows (where your wrappers change/fix things), and MKL, which by default has a G77 ABI AFAIU.

> Includes all the changes in [gh-18354](https://github.com/scipy/scipy/pull/18354).

Please try to rebase the commit from #18354 and include it here (or directly include a revert of https://github.com/h-vetinari/scipy/commit/c73e0883501a21cfaf7220a8118a6155857b5829). If you want I can do it.

h-vetinari commented:
> Just for clarity, the segfaults not only occurred on 32-bit machines, but also 64-bit ones, [...]

In particular, it's conceivable that we need to decide on the complex type based on the convention of the fortran compiler being used. Compare for example this [logic](https://github.com/flame/blis/blob/a72e4569f2a03cc3578c019bf7ce25491a44137d/configure#L4024-L4071) in blis.

h-vetinari commented:
> Please try to rebase the commit from #18354 and include it here (or directly include a revert of c73e0883501a21cfaf7220a8118a6155857b5829).

An empty commit is not really helpful (won't show up in git blame etc.). Please rebase it (or a fresh revert of c73e0883501a21cfaf7220a8118a6155857b5829) to the start of this PR and layer your changes on top.

thalassemia commented:
Yeah, I'm not sure how to do the rebase. Would be great if you could make that commit @h-vetinari!

rgommers commented:
Awesome! I'll review in more detail very soon. Just a quick note on complex types, in case you hadn't seen this yet (from https://numpy.org/devdocs/reference/c-api/coremath.html#support-for-complex-numbers):

<img width="471" alt="image" src="https://github.com/scipy/scipy/assets/98330/021599b7-bfe1-4d2d-a3ec-9105b2238a55">

thalassemia commented:
Oh that makes sense! I basically copied the preprocessor flags at the top of my wrappers from `npy_common.h`and was confused that using `npy_complex64/128` directly still caused segfaults. I didn't realize the change to C99 types only occurred in 2.0 and is not in any versions that we build with here.

rgommers commented:
> I didn't realize the change to C99 types only occurred in 2.0 and is not in any versions that we build with here.

We do test 2.0, but only in the two `Prerelease deps` jobs.

rgommers commented:
The one failure in the Linux aarch64 job isn't too concerning:
```
FAILED scipy/sparse/linalg/_eigen/tests/test_svds.py::Test_SVDS_PROPACK::test_svd_simple[csc_matrix-True-True-4-A1] - AssertionError: 
Not equal to tolerance rtol=1e-07, atol=3e-10

Mismatched elements: 2 / 16 (12.5%)
Max absolute difference: 6.35983951e-10
Max relative difference: 3.99680289e-15
 x: array([[ 1.000000e+00,  2.814834e-15,  2.269493e-12, -6.359840e-10],
       [ 2.814834e-15,  1.000000e+00,  5.029722e-14, -1.138435e-11],
       [ 2.269493e-12,  5.029722e-14,  1.000000e+00, -2.409678e-13],
       [-6.359840e-10, -1.138435e-11, -2.409678e-13,  1.000000e+00]])
 y: array([[1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [0., 0., 1., 0.],
       [0., 0., 0., 1.]]
```

The Windows build failure is unrelated. I've pushed a rebase, will see if we can get CI happy here.

rgommers committed:
TST: increase `atol` value is `svds` test for PROPACK only

Due to a failure on Linux aarch64, visible in the Cirrus CI job.

Also a very small bump in an ARPACK test to avoid a test failure
on macOS arm64 in CI.

[skip actions] [skip circle]

rgommers commented:
One more, for macOS arm64:
```
scipy/sparse/linalg/_eigen/tests/test_svds.py:311: in test_svds_parameter_tol
    assert error < accuracy
E   assert 2.087038755703438e-15 < 2e-15
```

rgommers commented on scipy/sparse/linalg/_eigen/tests/test_svds.py:
```diff
@@ -316,17 +304,15 @@ def err(tol):
             error = err(tol)
             assert error < accuracy
-            assert error > accuracy/10
+            # Why do we test a lower bound for error?
```
Agreed. that doesn't many too much sense. I'll remove it now that I'm touching this test anyway.

rgommers committed:
BLD: remove use of `-Wno-incompatible-pointer-types`

This will turn into a hard error in GCC 14, so we should no
longer suppress this warning for convenience.

rgommers reviewed (commented):
CI is happy now. The changes in `_build_utils/src` all look good to me, as does re-enabling PROPACK.

I am still looking at `_generate_pyx.py` a bit more carefully. I tried removing suppression of incompatible pointer warnings (see https://github.com/rgommers/scipy/commit/8647d653ac2e2024798d83ed2e8f8bbbbfb3b749, important to do soon because it'll break with GCC 14), but this PR didn't make a change there yet - still seeing warnings like these for all six wrapped functions:
```
[38/41] Compiling C object scipy/linalg/cython_l...linux-gnu.so.p/meson-generated_cython_lapack.c.o
scipy/linalg/cython_lapack.cpython-310-x86_64-linux-gnu.so.p/cython_lapack.c: In function '__pyx_f_5scipy_6linalg_13cython_lapack_cladiv':
scipy/linalg/cython_lapack.cpython-310-x86_64-linux-gnu.so.p/cython_lapack.c:4344:33: warning: passing argument 1 of 'cladivwrp_' from incompatible pointer type [-Wincompatible-pointer-types]
 4344 |   F_FUNC(cladivwrp, CLADIVWRP)((&__pyx_v_out), ((npy_complex64 *)__pyx_v_x), ((npy_complex64 *)__pyx_v_y));
      |                                ~^~~~~~~~~~~~~
      |                                 |
      |                                 __pyx_t_float_complex *
In file included from scipy/linalg/cython_lapack.cpython-310-x86_64-linux-gnu.so.p/cython_lapack.c:1112:
scipy/linalg/_lapack_subroutines.h:23:50: note: expected '_Complex float *' but argument is of type '__pyx_t_float_complex *'
   23 | void F_FUNC(cladivwrp, CLADIVWRP)(npy_complex64 *ret, npy_complex64 *x, npy_complex64 *y);
      |                                   ~~~~~~~~~~~~~~~^~~
scipy
```

In the generated `_blas_subroutines.h` the signatures use numpy types:
```C
void F_FUNC(zdotcwrp, ZDOTCWRP)(npy_complex128 *ret, int *n, npy_complex128 *zx, int *incx, npy_complex128 *zy, int *incy);
```
while the generated `cython_lapack.c` tries to use native types only for the first argument and numpy types for the third argument:
```C
  F_FUNC(zladivwrp, ZLADIVWRP)((&__pyx_v_out), ((npy_complex128 *)__pyx_v_x), ((npy_complex128 *)__pyx_v_y));
```
with `__pyx_v_out` being declared as:
```C
#if CYTHON_CCOMPLEX && (1) && (!0 || __cplusplus)
  #ifdef __cplusplus
    typedef ::std::complex< double > __pyx_t_double_complex;
  #else
    typedef double _Complex __pyx_t_double_complex;
  #endif
#else
    typedef struct { double real, imag; } __pyx_t_double_complex;
#endif
```

coming from `cython_lapack.pyx`:
```C
cdef extern from "_lapack_subroutines.h":
    void _fortran_zladiv "F_FUNC(zladivwrp, ZLADIVWRP)"(z *out, npy_complex128 *x, npy_complex128 *y) nogil
```

It's a bit of a messy puzzle, I'm not quite sure what the right solution is. What do you think @thalassemia?

h-vetinari commented on .gitmodules:
```diff
@@ -3,8 +3,9 @@
 [submodule "scipy/sparse/linalg/_propack/PROPACK"]
 	path = scipy/sparse/linalg/_propack/PROPACK
-	url = https://github.com/scipy/PROPACK
+	url = https://github.com/thalassemia/PROPACK.git
 	shallow = true
+	branch = wrapped_blas
```
Do we want to merge https://github.com/thalassemia/PROPACK/commit/cd4e3d3bfc7d1f667111076adc7bccb3e23d0a71 into https://github.com/scipy/PROPACK first?

thalassemia replied:
Yes, we should merge that and bump the submodule commit in this PR.

h-vetinari replied:
Open PR for xref: https://github.com/scipy/PROPACK/pull/7

h-vetinari commented on scipy/_build_utils/src/wrap_dummy_g77_abi.c:
So in the meson file we do:
```
if use_g77_abi
  g77_abi_wrappers = declare_dependency(
    sources: ['_build_utils/src/wrap_g77_abi.c'],
    ...
  )
else
  g77_abi_wrappers = declare_dependency(
    sources: ['_build_utils/src/wrap_dummy_g77_abi.c'],
    ...
  )
```

I think "dummy_g77" is not a great name for "not the g77 ABI". I realize it was named like that before, but thinking out loud if we can have better naming here.

thalassemia replied:
I'm ambivalent about the naming. On one hand, the `dummy` moniker makes sense because the file only exists to be a mock G77 wrapper when the real one is not needed. On the other hand, this interpretation of the name is not obvious at first glance, as you pointed out.

h-vetinari commented on scipy/_build_utils/src/wrap_g77_abi.c:
I also think both `wrap_*` files should have a prominent comment explaining that:
* why they exist (create compatibility layer for complex-handling between different compilers/runtimes)
* what the alternative is (g77 vs non-g77)
* what distinguishes the current file from the alternative implementation.

thalassemia replied:
I just added a brief docstring to the top of the two wrapper files.

h-vetinari replied:
Great, thanks!

h-vetinari reviewed (commented):
Overall this is excellent and amazing to see! :)

Just some high-level comments about naming & documentation, to keep this hopefully more easily maintainable.

thalassemia committed:
BUG: Cast to resolve incompatible pointer warnings

thalassemia commented:
Thanks for pointing out those warnings and investigating @rgommers! On my machine, casting the return type variable to the expected Numpy type gets rid of those warnings (3ccf67c).

thalassemia committed:
DOC: Docstring for G77 wrapper files

rgommers commented:
> Thanks for pointing out those warnings and investigating @rgommers! On my machine, casting the return type variable to the expected Numpy type gets rid of those warnings ([3ccf67c](https://github.com/scipy/scipy/commit/3ccf67cddc9f8816804abee16b71a54c29e2d286)).

Thanks! Works for me too, and the generated code looks better. Would you mind including my commit to disable `-Wno-incompatible-pointer-types` in this PR? With that and the update to the PROPACK module, this looks ready to merge.

thalassemia committed:
REV: Undo 65050cc to use main PROPACK repo

thalassemia committed:
ENH: Update propack submodule to remove G77 code

thalassemia commented:
> Note the use of scipy-openblas wheels requires a `scipy_` prefix, which in turn requires something like defining `BLAS_SYMBOL_PREFIX=scipy_` (done by the pkgconfig file in the wheel) and using the `BLAS_FUNC` and `CBLAS_FUNC` macros.

Thanks for pointing this out @mattip! In #19816, I added a script to generate C wrappers that route all BLAS/LAPACK function calls to their properly prefixed and/or suffixed names. I plan on iterating on that PR once this one is merged.

rgommers reviewed (approved):
Looks ready and CI is happy, so in it goes. Thanks a lot @thalassemia. And thanks @h-vetinari for the review.
